[송영섭] sprint 8

songyoungsub opened the pull request:
## 기본 요구사항
### 공통

[0] Github에 스프린트 미션 PR을 만들어 주세요.

[0] Next.js를 사용해 진행합니다.

### 자유 게시판 페이지

[0] 게시글 목록에서 드롭다운을 사용하여 "최신 순"으로 정렬할 수 있도록 합니다.

[0] 본인이 이전 미션에서 생성한 게시글 목록 조회 API를 활용해 GET 메서드로 데이터를 가져옵니다.

[0] 게시글 제목에 검색어가 일부 포함되면 해당 게시글을 검색할 수 있도록 합니다.

[0] 이미지는 디폴트 이미지로 프론트엔드에서 처리해 주세요.

[0] 게시글 닉네임 및 좋아요 개수 역시 임의값으로 프론트엔드에서 처리해주세요.

[0] 베스트 게시글은 최신순 3개 게시글을 요청으로 데이터를 가져와 구현해주세요.

[0] 자유게시판 페이지에서 특정 게시글을 클릭하면 해당 게시물의 상세 페이지로 이동합니다.

### 게시글 등록 & 수정 페이지

[0] 각 input 필드에 정확한 placeholder 값을 입력합니다.

[0] 모든 input 필드에 값을 입력하면 '등록' 버튼이 활성화됩니다.

[0] 본인이 이전 미션에서 생성한 게시글 생성 API를 활용해 POST 메서드로 게시글을 등록합니다.

[0] '등록' 버튼을 누르면 해당 게시물 상세 페이지로 이동합니다.

[0] 게시글 수정 페이지 UI는 게시글 등록 페이지와 동일합니다.

[0] 본인이 이전 미션에서 생성한 게시글 상세 API의 PATCH 메소드를 사용하여 게시물을 수정합니다.

### 게시글 상세 페이지

[0] 본인이 이전 미션에서 생성한 게시글 상세 API의 GET 메소드를 사용하여 데이터를 가져옵니다.

[0] 본인이 이전 미션에서 생성한 게시글 상세 API의 DELETE 메소드를 사용하여 게시물을 삭제합니다.

[0] 댓글 input에 값을 입력하면 '등록' 버튼이 활성화됩니다.

[0] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 POST 메소드로 댓글을 등록합니다.

[0] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 PATCH 메소드로 댓글을 수정합니다.

[0] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 DELETE 메소드로 댓글을 삭제합니다.

## 심화 요구사항
### 공통

[0] 디자인 시안에 따라 반응형 디자인을 구현합니다.

[x] (생략 가능) 원한다면 지금까지 진행한 모든 React 코드를 Next.js로 마이그레이션 해주세요.

마이그레이션에 상당한 시간이 소요될 수 있으므로 진행을 권장하지 않습니다.

## 강사님께
오류 해결했습니다
주소 : https://1-sprint-mission-fe.vercel.app/

Commit:
feat: 자유게시판 틀 작성 및 Header와 Footer 이식

Commit:
feat: 글쓰기 페이지 틀 작성

Commit:
feat: 상페이지 틀 생성

Commit:
style: 베스트 게시글 css 작성
BestArticle.js, BestArticleBody.js의 css 작성

Commit:
style: 게시판 디자인 작성
Article.js와 ArticleBody.js의 css 작성 및 몇몇 중복 css 수정

Commit:
style: 일부 css 이름 변경

Commit:
style: 게시글 쓰기 디자인 작성
PostForm.js 의 css 작성

Commit:
style: 상세페이지 디자인 작성
CommentFrom.js, CommentList.js, CommentListBody.js, ParticularInformation.js의 css 작성
참고: 아직 반응형 디자인은 작성하지 않았으며 기능 구현 후 할 예정

Commit:
style: 일부 css 값 수정

Commit:
feat: 게시글 수정 페이지 생성
기존 게시글 등록 페이지를 복사 및 일부 수정, css도 게시글 등록 페이지 사용

Commit:
feat: 현재 api 적용 중

Commit:
feat: 댓글 수정 및 무한 스크롤 제외 대부분 기능 구현

Commit:
feat: 무한 스크롤 구현

Commit:
feat: 댓글 수정 기능 구현

Commit:
feat: 일부 오류 수정
댓글 수정을 누르고 댓글을 삭제 시 수정 form이 초기화 되지 않는 오류 수정

Commit:
style: 자유게시판 반응형 디자인 구현

Commit:
style: 게시물 생성 및 수정 페이지 반응형 디자인 구현

wildCodingWarrior commented on next-mission/components/freeNoticeBoard/Article.module.css:
```diff
@@ -0,0 +1,107 @@
+  margin-top: 40px;
+  margin-bottom: 293px;
+}
+
```
BEM 컨벤션 따르려고 해보신건가요? 
module.css 쓰는 만큼 이번에는 모든 css selector를 camelCase로 한번 해봅시다.

wildCodingWarrior commented on next-mission/components/freeNoticeBoard/BestArticle.js:
```diff
@@ -0,0 +1,40 @@
+import BestArticleBody from "./BestArticleBody";
+import style from "./BestArticle.module.css";
+
+export default function BestArticle({ list }) {
+  if (!list[0]) return;
+
+  const [idx, setIdx] = useState([0, 1, 2]);
+
+  // 스크린 크기에 따른 로고 변경
+  const handleResize = useCallback(() => {
+    const length = window.innerWidth;
+
+    if (length >= 1200) {
+      setIdx([0, 1, 2]);
+    } else if (length >= 768 && length < 1200) {
+      setIdx([0, 1]);
+    } else if (length >= 375 && length < 768) {
+      setIdx([0]);
+    }
+  }, []);
+
+  useResize(handleResize);
+
+  return (
+    <div className={style.BestArticle_contaner}>
+      <div className={style.BestArticle_title}>베스트 게시글</div>
+      <ul className={style.BestArticle_ul}>
+        {idx.map((idx) => {
+          return (
+            <li className={style.BestArticle_li}>
+              <BestArticleBody list={list[idx]} />
+            </li>
+          );
+        })}
+      </ul>
+    </div>
+  );
+}
```
여기서 if (!list[0]) return; 이 코드의 경우 hooks 룰에 어긋나는 부분이 있습니다.
그것은 바로 아래 있는 useState와 useCallback이 조건적으로 실행이 된다는 점입니다. render하고 싶은 것이 없을 경우에는 부모 컴포넌트에서 렌더링을 막던가, 아니면 JSX 부분의 return 바로 위에서 해주세요.

wildCodingWarrior commented on next-mission/components/freeNoticeBoard/BestArticle.js:
```diff
@@ -0,0 +1,40 @@
+      <ul className={style.BestArticle_ul}>
+        {idx.map((idx) => {
+          return (
+            <li className={style.BestArticle_li}>
```
key가 없네요!

wildCodingWarrior commented on next-mission/components/particularPage/CommentList.js:
```diff
@@ -0,0 +1,53 @@
+
+export default function CommentList() {
+  const [list, setList] = useState([1, 2, 3, 4, 5]); // 예정
+  const [noList, setNoList] = useState(false);
```
noList를 쓰는 것 보다는 list.length를 이용하는게 더 좋겠습니다.

wildCodingWarrior commented on next-mission/components/particularPage/ParticularInformation.js:
```diff
@@ -0,0 +1,72 @@
+
+export default function ParticularInformation() {
+  const [hideDropDown, setHideDropDown] = useState(false);
+
```
조건부 렌더링을 이렇게 개선하면 어떨까요? {hideDropDown || (...)} 대신 {showDropDown && (...)}로요.

wildCodingWarrior commented on next-mission/pages/freeNoticeBoard/index.js:
```diff
@@ -0,0 +1,29 @@
+  }
+}
+
+export default function freeNoticeBoard({ total, list}) {
```
FreeNoticeBoard로 수정합시다.

wildCodingWarrior commented on next-mission/pages/freeNoticeBoard/[id].js:
```diff
@@ -0,0 +1,17 @@
+import ParticularInformation from "@/components/particularPage/ParticularInformation";
+import Head from "next/head";
+
+export default function particularPage() {
```
이것도 ParticularPage로 수정합시다.

wildCodingWarrior commented on next-mission/lib/axios.js:
```diff
@@ -0,0 +1,7 @@
+import axios from "axios";
+
+const instance = axios.create({
+    baseURL:'https://one-sprint-mission-be-5h16.onrender.com'
```
env에 넣으면 더 좋겠죠?

wildCodingWarrior reviewed (commented):
고생하셨습니다. 일요일까지 잘 마무리해봅시다.

Commit:
style: 상세 페이지 반응형 디자인 구현

Commit:
style: css 이름 카멜방식으로 변경

Commit:
feat: 리뷰받은 걸 기준으로 수정

Commit:
feat: axios URL를 env 파일에 분리 및 기존 다른 페이지 파일 등에 쓰는 axios 변수명 변경

Commit:
feat: 게시글 수정 페이지 css 이름 변경 및 스피너 컴포 추가

Commit:
feat: 배포를 위한 URL 작성

Commit:
feat: 배포를 위한 오류 제거

Commit:
feat: URL 변경

Commit:
spinner 컴포가 문제로 예상 테스트를 위해 삭제

Commit:
URL 삽입

Commit:
테스트 1

Commit:
feat: 오류 수정
서버사이드 랜더링에서 받아온 값 중 배열을 인덱스로 확인하여 html을 보여줄지 말지 정하는 것에서 오류가 발생, length 메소드를 이용하여 해결

Commit:
feat: URL 제거

wildCodingWarrior commented on next-mission/components/particularPage/CommentList.js:
```diff
@@ -33,17 +32,15 @@ export default function CommentList({
   return (
     <>
-      {noList && (
+      {comment[0] || (
```
앗 제 리뷰가 전달이 이상하게 된 것 같은데..

{comment[0] || ()}

이런식으로 표현하면 의도가 명확하지 않을 수 있다고 생각합니다. 첫번째 댓글만 보여주려는건지, 댓글이 있는지 확인하려는 것인지 헷갈리구요.
그리고 || 연산자 자체가 falsy한 값에 대해서 예상치 못한 동작을 할 수 도 있구요.

그래서 저는 권장하는 건 {comment.length === 0 && ()} 이런 형식입니다. 더 의도가 명확하거든요.

wildCodingWarrior reviewed (approved):
지적한 부분 아주 잘 바꾸셨네요! 그 외에는 API만 좀 나누면 좋을 것 같긴 합니다.
